samples/net/sockets/http_server/src/main.c

```diff
@@ -37,59 +44,139 @@ struct http_resource_detail_static index_html_gz_resource_detail = {
+		return -1;
+	}
+	}
+}
+
+static struct http_resource_detail_dynamic uptime_resource_detail = {
+	.common = {
+			.type = HTTP_RESOURCE_TYPE_DYNAMIC,
+			.bitmask_of_supported_http_methods = BIT(HTTP_GET),
+		},
+	.cb = uptime_handler,
+	.data_buffer = uptime_buf,
+	.data_buffer_len = sizeof(uptime_buf),
+	.user_data = NULL,
+};
+
+static void led_control(bool on)
+{
+	LOG_INF("POST request turning LED %s", on ? "on" : "off");
+
+	if (leds_dev != NULL) {
+		if (on) {
+			led_on(leds_dev, 0);
+		} else {
+			led_off(leds_dev, 0);
+		}
+	}
+}
+
+static void parse_led_post(uint8_t *buf, size_t len)
+{
+	if ((len == sizeof("led=on") - 1) && (0 == memcmp("led=on", buf, len))) {
+		led_control(true);
+	} else if ((len == sizeof("led=off") - 1) && (0 == memcmp("led=off", buf, len))) {
+		led_control(false);
+	} else {
+		LOG_WRN("Unexpected POST payload");
+	}
+}
+
+static int led_handler(struct http_client_ctx *client, enum http_data_status status,
 		       uint8_t *buffer, size_t len, void *user_data)
 {
-#define MAX_TEMP_PRINT_LEN 32
-	static char print_str[MAX_TEMP_PRINT_LEN];
-	enum http_method method = client->method;
-	static size_t processed;
+	static uint8_t post_payload_buf[32];
+	static size_t cursor;
 
-	__ASSERT_NO_MSG(buffer != NULL);
+	LOG_DBG("LED handler status %d, size %zu", status, len);
 
 	if (status == HTTP_SERVER_DATA_ABORTED) {
-		LOG_DBG("Transaction aborted after %zd bytes.", processed);
-		processed = 0;
+		cursor = 0;
 		return 0;
 	}
 
-	processed += len;
+	if (len + cursor > sizeof(post_payload_buf)) {
+		cursor = 0;
+		return -ENOMEM;
+	}
 
-	snprintf(print_str, sizeof(print_str), "%s received (%zd bytes)",
-		 http_method_str(method), len);
-	LOG_HEXDUMP_DBG(buffer, len, print_str);
+	/* Copy payload to our buffer. Note that even for a small payload, it may arrive split into
+	 * chunks (e.g. if the header size was such that the whole HTTP request exceeds the size of
+	 * the client buffer).
+	 */
+	memcpy(post_payload_buf + cursor, buffer, len);
+	cursor += len;
 
 	if (status == HTTP_SERVER_DATA_FINAL) {
-		LOG_DBG("All data received (%zd bytes).", processed);
-		processed = 0;
+		parse_led_post(post_payload_buf, cursor);
+		cursor = 0;
 	}
 
-	/* This will echo data back to client as the buffer and recv_buffer
-	 * point to same area.
-	 */
-	return len;
+	return 0;
 }
 
-struct http_resource_detail_dynamic dyn_resource_detail = {
+static struct http_resource_detail_dynamic led_resource_detail = {
 	.common = {
-		.type = HTTP_RESOURCE_TYPE_DYNAMIC,
-		.bitmask_of_supported_http_methods =
-			BIT(HTTP_GET) | BIT(HTTP_POST),
-	},
-	.cb = dyn_handler,
-	.data_buffer = recv_buffer,
-	.data_buffer_len = sizeof(recv_buffer),
+			.type = HTTP_RESOURCE_TYPE_DYNAMIC,
+			.bitmask_of_supported_http_methods = BIT(HTTP_POST),
+		},
+	.cb = led_handler,
+	.data_buffer = led_buf,
+	.data_buffer_len = sizeof(led_buf),
 	.user_data = NULL,
 };
 
-HTTP_RESOURCE_DEFINE(dyn_resource, test_http_service, "/dynamic",
```

Comment:
Can we keep the old "dynamic" resource in the sample? Maybe rename it to "echo" or something to avoid confusion. It may indeed not be the best demo, but it was actually pretty useful for testing large transfers, would be good to have something like this at hand.

Reply:
Added back in as `echo_resource`